### PR TITLE
add rayon feature

### PR DESCRIPTION
**Pull request by oligamiq**

### CODE REVIEW

File: Cargo.toml
Comment (fintelia) [inline Cargo.toml:83]: This isn't needed anymore, since rustdoc [recently added it as the default](https://github.com/rust-lang/docs.rs/pull/2390/files).

**Comment by fintelia** [comment]
Not a maintainer, but FYI this is an API breaking change because any downstream code relying on the now cfg'd out methods will no longer compile. If rayon was left as a default feature, this change could be done in a backwards compatible way

**Comment by oligamiq** [comment]
All the tests have passed, but indeed, this might be a breaking change.

**Comment by johannesvollmer** [comment]
Thanks for taking the time to contribute! The changed you proposed seem solid. I would have solved it differently, but I don't have the time to do that:

As the calls to rayon are quite few and isolated, it should be possible to keep the API as-is. Then at runtime falling back to sequential processing if the rayon feature is disabled. This would require some more complicated code changed, but it would avoid us having to sprinkle `cfg` attributes in multiple unrelated code sections, which seems hard to maintain in my eyes (and is not backwards compatible).

The backwards incompatible changes would require major version upgrade, `exr 2.0`, which makes it harder for people to use the newest versions, as they have to manually update the version in their projects.

What do you think?

**Comment by oligamiq** [comment]
I think it’s possible to maintain all the public structs and APIs while marking them as deprecated and falling back. I'll be busy for a few days, but after that, I can give it a try if you'd like.

**Comment by johannesvollmer** [comment]
Checked the current code and found something problematic, there is at least one public function that exposes a rayon item at compile time: `pub fn new_with_thread_pool`. Someone might rely on this function, it references the rayon thread pool in the function type signature. This means we cannot get around a breaking change. Please correct me if I'm wrong.

**Comment by oligamiq** [comment]
That's right. I share the same opinion. While it's possible to replace the dependent types with generics, it would cause some significant breaking changes. In the case of removing Rayon from image-rs, it's acceptable to have Rayon set to default ON. However, setting Rayon to default ON is generally not advisable.

**Comment by oligamiq** [comment]
Since there seem to be a lot of `issues` left, how about removing rayon from the default feature list at the time of a major version update?

**Comment by johannesvollmer** [comment]
after thinking about this, I'm happy to merge this. (of course we need to fix the 3 small conflicts, but that's not a big deal.) thanks for your contribution!

File: .github/workflows/rust.yml
Comment (johannesvollmer) [inline .github/workflows/rust.yml:26]: we might need to invert this to run without default features since rayon will be enabled by default

File: Cargo.toml
Comment (johannesvollmer) [inline Cargo.toml:36]: flume has been removed, just remove this line to solve the conflict

**Comment by johannesvollmer** [comment]
@fintelia what's the thread pool choice of the image crate? i'm assuming the rayon feature will be turned off in the image crate. that results in 50%-80% loss in performance for exrs. but I think you still have threading in the image crate?

maybe exrs can adapt to other parallelism crates than rayon? or maybe std has a thread pool by now and I didn't notice?

**Comment by fintelia** [comment]
Yes. After this lands, `image` will switch to only enabling the rayon feature on `exr` when its own rayon feature is enabled

**Comment by oligamiq** [comment]
I accepted the CI-related reviews and resolved the conflicts.

**Review by johannesvollmer — APPROVED** [review]
Thanks for taking the time for the follow up! I appreciate your contributions.

**Pull request merged**
